File: android/src/main/kotlin/io/alexrintt/sharedstorage/deprecated/lib/DocumentFileColumn.kt
package io.alexrintt.sharedstorage.deprecated.lib

import android.database.Cursor
import android.provider.DocumentsContract
import java.lang.NullPointerException

private const val PREFIX = "DocumentFileColumn"

enum class DocumentFileColumn {
  ID,
  DISPLAY_NAME,
  MIME_TYPE,
  SUMMARY,
  LAST_MODIFIED,
  SIZE
}

enum class DocumentFileColumnType {
  LONG,
  STRING,
  INT
}

fun parseDocumentFileColumn(column: String): DocumentFileColumn? {
  val values = mapOf(
    "$PREFIX.COLUMN_DOCUMENT_ID" to DocumentFileColumn.ID,
    "$PREFIX.COLUMN_DISPLAY_NAME" to DocumentFileColumn.DISPLAY_NAME,
    "$PREFIX.COLUMN_MIME_TYPE" to DocumentFileColumn.MIME_TYPE,
    "$PREFIX.COLUMN_SIZE" to DocumentFileColumn.SIZE,
    "$PREFIX.COLUMN_SUMMARY" to DocumentFileColumn.SUMMARY,
    "$PREFIX.COLUMN_LAST_MODIFIED" to DocumentFileColumn.LAST_MODIFIED
  )

  return values[column]
}

fun documentFileColumnToRawString(column: DocumentFileColumn): String? {
  val values = mapOf(
    DocumentFileColumn.ID to "$PREFIX.COLUMN_DOCUMENT_ID",
    DocumentFileColumn.DISPLAY_NAME to "$PREFIX.COLUMN_DISPLAY_NAME",
    DocumentFileColumn.MIME_TYPE to "$PREFIX.COLUMN_MIME_TYPE",
    DocumentFileColumn.SIZE to "$PREFIX.COLUMN_SIZE",
    DocumentFileColumn.SUMMARY to "$PREFIX.COLUMN_SUMMARY",
    DocumentFileColumn.LAST_MODIFIED to "$PREFIX.COLUMN_LAST_MODIFIED"
  )

  return values[column]
}

fun parseDocumentFileColumn(column: DocumentFileColumn): String {
  val values = mapOf(
    DocumentFileColumn.ID to DocumentsContract.Document.COLUMN_DOCUMENT_ID,
    DocumentFileColumn.DISPLAY_NAME to DocumentsContract.Document.COLUMN_DISPLAY_NAME,
    DocumentFileColumn.MIME_TYPE to DocumentsContract.Document.COLUMN_MIME_TYPE,
    DocumentFileColumn.SIZE to DocumentsContract.Document.COLUMN_SIZE,
    DocumentFileColumn.SUMMARY to DocumentsContract.Document.COLUMN_SUMMARY,
    DocumentFileColumn.LAST_MODIFIED to DocumentsContract.Document.COLUMN_LAST_MODIFIED
  )

  return values[column]!!
}

/// `column` must be a constant String from `DocumentsContract.Document.COLUMN*`
fun typeOfColumn(column: String): DocumentFileColumnType? {
  val values = mapOf(
    DocumentsContract.Document.COLUMN_DOCUMENT_ID to DocumentFileColumnType.STRING,
    DocumentsContract.Document.COLUMN_DISPLAY_NAME to DocumentFileColumnType.STRING,
    DocumentsContract.Document.COLUMN_MIME_TYPE to DocumentFileColumnType.STRING,
    DocumentsContract.Document.COLUMN_SIZE to DocumentFileColumnType.LONG,
    DocumentsContract.Document.COLUMN_SUMMARY to DocumentFileColumnType.STRING,
    DocumentsContract.Document.COLUMN_LAST_MODIFIED to DocumentFileColumnType.LONG,
    DocumentsContract.Document.COLUMN_FLAGS to DocumentFileColumnType.INT
  )

  return values[column]
}

fun cursorHandlerOf(type: DocumentFileColumnType): (Cursor, Int) -> Any? {
  when (type) {
    DocumentFileColumnType.LONG -> {
      return { cursor, index ->
        try {
          cursor.getLong(index)
        } catch (e: NullPointerException) {
          null
        }
      }
    }
    DocumentFileColumnType.STRING -> {
      return { cursor, index ->
        try {
          cursor.getString(index)
        } catch (e: NullPointerException) {
          null
        }
      }
    }
    DocumentFileColumnType.INT -> {
      return { cursor, index ->
        try {
          cursor.getInt(index)
        } catch (e: NullPointerException) {
          null
        }
      }
    }
  }
}

File: lib/src/saf/models/document_file_column.dart
// /// Representation of the available columns of `DocumentsContract.Document.<Column>`
// ///
// /// [Refer to details](https://developer.android.com/reference/android/provider/DocumentsContract.Document)
// enum DocumentFileColumn {
//   /// Equivalent to [`COLUMN_DOCUMENT_ID`](https://developer.android.com/reference/android/provider/DocumentsContract.Document#COLUMN_DOCUMENT_ID)
//   id('COLUMN_DOCUMENT_ID'),

//   /// Equivalent to [`COLUMN_DISPLAY_NAME`](https://developer.android.com/reference/android/provider/DocumentsContract.Document#COLUMN_DISPLAY_NAME)
//   displayName('COLUMN_DISPLAY_NAME'),

//   /// Equivalent to [`COLUMN_MIME_TYPE`](https://developer.android.com/reference/android/provider/DocumentsContract.Document#COLUMN_MIME_TYPE)
//   mimeType('COLUMN_MIME_TYPE'),

//   /// Equivalent to [`COLUMN_LAST_MODIFIED`](https://developer.android.com/reference/android/provider/DocumentsContract.Document#COLUMN_LAST_MODIFIED)
//   lastModified('COLUMN_LAST_MODIFIED'),

//   /// Equivalent to [`COLUMN_SIZE`](https://developer.android.com/reference/android/provider/DocumentsContract.Document#COLUMN_SIZE)
//   size('COLUMN_SIZE'),

//   /// Equivalent to [`COLUMN_SUMMARY`](https://developer.android.com/reference/android/provider/DocumentsContract.Document#COLUMN_SUMMARY)
//   summary('COLUMN_SUMMARY');

//   const DocumentFileColumn(this.androidEnumItemName);

//   static const String _kAndroidEnumTypeName = 'DocumentFileColumn';

//   final String androidEnumItemName;

//   String get androidEnumItemId => '$_kAndroidEnumTypeName.$androidEnumItemName';

//   @override
//   String toString() => androidEnumItemId;
// }
